readme
Welcome
-------
This is YWCMAIL 0.96. This is an essential upgrade to the 0.95 version. Please see the details of the changes in the file 000-Readme.txt. The main upgrade is a simple access list and handling of more complex MIME messages.

What YWCMAIL is?
----------------
First, thanks for your support to ywcmail. Actually, ywcmail is a simple and
fast webmail CGI client written on pure C, even more, it is a SECURE
SOLUTION for pop3 protocol. Hackers can easily broke into port 110 and
get your e-mail, and this means that you have no PIRACY. OK, you may ask why 
don't you use SSL+POP3? I can tell you the settings are complex enough to 
take you more than 3 hours. 

If you use ywcmail, you can get a SECURE POP3 and a SECURE webmail faster.

Ywcmail gets your e-mail though port 110 on the localhost, handle it and
finally report to the users though web interface (Port80). This means you
can close port110 of your firewall. And at the same time, you can get a webmail.
Don't forget, the whole procedure need to compile ywcmail and make it runs 
need only 10 MINUTES. 

HOWEVER, you should put ywcmail under Apache+SSL (port443) to protect your
data. This is RECOMMENDED 

Installation 
------------
You need the followings:

- Apache 
- GCC compiler (or standard C compiler)


Steps are as follows:

1. Open the file 'config.h' with a text editor and make changes to it.
   Some of the available options are; default host names, language, colors,
   frames/no frames and return address host name.  Odds are, you won't have to
   change any of these for  ywcmail to work, however, some changes (like
   changing the default host names) may make ywcmail easier to use. If you want 
   to change the layout of the HTML code, edit main.c

2. Type 'make'.  This will make the file 'webmail.cgi'.  

3. Copy this file to a CGI enabled directory in your web server.  Additionally, 
   you will probably want to copy the 'webmail' subdirectory to the root of 
   your web server's htdocs directory.

4. DONE. This takes you <10 mintues to reach here. Enjoy it ! :D  

-----------
Chan Yin Chi  < 0330@bigfoot.com >
http://ywcmail.sunsite.dk

changelog
New in 0.95 (released on 20/01/03)
-----------
- fixed some small bugs 
- updated translation (traditional chinese)
- some new CSS effects

New in 0.94
-----------
- new style sheet 
- new interface
- new Icons

New in 0.93  (seems that 1.0 is closer now :))
-----------
- new translation (simplified chinese)
- code cleanup
- cleanup the HTML code in display

New in 0.92
-----------
- fixed error code in logout
- fixed MIME bug


New in 0.91
-----------
- new translation (traditional chinese)
- Imporve MIME 

New in 0.9
----------
- minor bugfix
  - after entering wrong password, the images can't be display correctly. 
     - (Thanks to Percy Cheung for discovering the bug.)

- text colour changed  




main.h
/*
    ywcmail -- simple pop3/smtp web e-mail cgi
    Copyright (C) 2002  Chan Yin Chi < 0330@bigfoot.com >
    code based on null webmail
    Additional Code Copyright (C) 2016 < kt3@gmx.co.uk>
*/

#include <ctype.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/types.h>
#ifdef WIN32
	#pragma comment(lib, "wsock32.lib")
	#include <winsock.h>
	#include <io.h>
	#include <sys/timeb.h>
	#define snprintf _snprintf
	#define vsnprintf _vsnprintf
	#define strcasecmp stricmp
	#define strncasecmp strnicmp
#else
	#include <dirent.h>
	#include <netdb.h>
	#include <paths.h>
	#include <syslog.h>
	#include <unistd.h>
	#include <netinet/in.h>
	#include <sys/resource.h>
	#include <sys/socket.h>
	#include <sys/time.h>
	#include <sys/wait.h>
#endif
#include "config.h"

#define RFC1123FMT "%a, %d %b %Y %H:%M:%S GMT"

typedef struct {
	char From[128];
	char Replyto[128];
	char To[512];
 	char Date[128];
	char Subject[128];
	char CC[1024];
	char contenttype[256];
	char boundary[128];
	char encoding[128];
} wmheader;
typedef struct {
	char ClientIP[16];
	char Connection[128];
	int  ContentLength;
	char ContentType[128];
	char Cookie[1024];
	char Host[64];
	char PathInfo[128];
	char *PostData;
	char QueryString[1024];
	char Referer[128];
	char RequestMethod[8];
	char RequestURI[1024];
	char ScriptName[128];
	char UserAgent[128];
} request_struct;
request_struct request;
extern char **environ;
int  connected;
char wmusername[64];
char wmpassword[64];
char wmservertype[6];
char wmpop3server[64];
char wmsmtpserver[64];
int  wmsocket;

#ifdef SUPPORT_IMAP
typedef struct {
        int  wmimapidx;
        char wmimapfolder[48];
        char wmimapfolderlist[64][48];
} wmimapsession_struct;
wmimapsession_struct wmimapsession;
#endif


/* main.c stuff */
void printheader(void);
/* cgi-lib.c functions */
int  IntFromHex(char *pChars);
void URLDecode(unsigned char *pEncoded);
char *str2html(char *instring);
void cgi_readenv(void);
void cgi_readpost(void);
char *getgetenv(char *query);
char *getmimeenv(char *query);
char *getpostenv(char *query);
char *get_mime_type(char *name);
char *strcasestr(const char *src, const char *query);
void striprn(char *string);
/* wmmain.c functions */
int EncodeBase64(char *src, int srclen);
int  DecodeBase64(char *src, char *ctype);
int  wmprintf(const char *format, ...);
int  wmfgets(char *buffer, int max, int fd);
void wmclose(void);
void wmexit(void);
void send_header(int cacheable, int status, char *title, char *extra_header, char *mime_type, int length, time_t mod);
void wmcookieget(void);
int  wmcookieset(void);
void wmcookiekill(void);
void webmailread(void);
void webmailraw(void);
void webmailfiledl(void);
void webmailwrite(void);
void webmaillist(void);
void webmailsend(void);
void webmaildelete(void);
/* wmserver.c functions */
int  wmserver_connect(void);
void wmserver_disconnect(void);
int  wmserver_smtpconnect(void);
int  wmserver_smtpauth(void);
void wmserver_smtpdisconnect(void);
int  wmserver_count(void);
int  wmserver_msgdele(int message);
int  wmserver_msghead(int message);
int  wmserver_msgretr(int message);
int  wmserver_msgsize(int message);
int  wmserver_uidl(int message, char *uidl);
int is_msg_end(char *buffer);

inline void logdata(const char* type, const char* source_file, const char* method, const char *format, ...);
//#define LOGGING(...)
#define LOGGING(...) do { logdata( "LOG", c_file_name, c_func_name, __VA_ARGS__); } while (0)
//#define LOGGING_DISPLAY(...)
#define LOGGING_DISPLAY(...) do { logdata( "DISP", c_file_name, c_func_name, __VA_ARGS__); } while (0)
//#define LOGGING_DISPLAY(...)
#define LOGGING_TRACE(...) do { logdata( "TRAC", c_file_name, c_func_name, __VA_ARGS__); } while (0)
//#define LOGGING_DISPLAY(...)
#define LOGGING_WMGETS(...) do { logdata( "GET", c_file_name, c_func_name, __VA_ARGS__); } while (0)


files.c
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include "config.h"

int test_path(char path[])
{
    struct stat st;
    if(stat(path,&st) == 0)
    {
        return 0;
    } else
    {
        return 1;
    }
}

FILE* open_file(char f[])
{
  FILE* fp = fopen(f, "r");
  if (fp == NULL) {
    return NULL;
  }
  return fp;
}

int is_user_in_file(FILE* fp,char user[]) 
{
  char line[1024] = {};
  while (NULL != fgets( line, 1024, fp) )
  {
    if (line[0] != '#')
    {
        if (strncmp( line, user, strlen(user) ) == 0)
        {
          return 0;	
        }
    }
  }
  return 1;
}

// Filename is wmsuername_key 
int wm_is_user_allowed(char wmusername[])
{
  int is_path = 0;
  is_path = test_path(ALLOWED_USERS);
  if (is_path ==0) 
  {
    FILE* fp = open_file(ALLOWED_USERS); // return filehandle or NULL
    if (fp != NULL)
    {
      return is_user_in_file(fp, wmusername);
    }
  }
  return 1;  // path does not exist
}

#ifdef BUILD_FOR_TESTS

int main_read_allow()
{
  if(read_is_user_allowed( "anon" ) == 0)
  {
     // "found"
  }

  return 0;
}

#endif


logging.c
/*
  Logging routine
  Copyright (C) 2016 < kt3@gmx.co.uk>

*/

#include "config.h"
#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdarg.h>
#include <stdlib.h>
#include <time.h>
#ifdef USE_SYSLOG
  #include <syslog.h>
#endif

inline void logdata(const char* type, const char* source_file, const char* method, const char *format, ...)
{
	char logbuffer[1024];
        char file[200];
        va_list ap;
        FILE *fp;

	pid_t pid;
	pid = getpid();

	//char msgdate[100];

        snprintf(file, sizeof(file)-1, LOG_FILENAME);

	fp=fopen(file, "a");
	if (fp!=NULL) {
        	va_start(ap, format);
                vsnprintf(logbuffer, sizeof(logbuffer)-1, format, ap);
               	va_end(ap);
                fprintf(fp, "%d: %22.22s, %6.6s, %7.7s, %10.10s, %20.20s: %s\n", pid, "May 2019", "unknown", type, source_file, method, logbuffer);
                fclose(fp);
	}
        else
        {
		#ifdef USE_SYSLOG
                	syslog( LOG_MAIL|LOG_INFO, "Could not open log file %s, to write msg %s", LOG_FILENAME, logbuffer );
	    	#endif
        }
}


i18n/strings-en.h
/*
    ywcmail -- simple pop3/smtp web e-mail cgi
    Copyright (C) 2002 Chan Yin Chi < 0330@bigfoot.com >
    Translation by Yourname <you@somewhere.com>

*/
#define ERR_NOUSER	"User name and password cannot be blank"
#define ERR_NOHOST	"Missing server names"
#define ERR_NOWINSOCK	"Winsock initialization failed"
#define ERR_DNS_POP3	"Could not find POP3 server"
#define ERR_DNS_SMTP	"Could not find SMTP server"
#define ERR_CON_POP3	"Could not connect to POP3 server"
#define ERR_CON_SMTP	"Could not connect to SMTP server"
#define ERR_NORECIPIENT	"No recipient specified - Message was not sent"
#define ERR_BIGPOST	"Bad Request - POST too large"
#define ERR_BADURI	"Bad Request - Malformed URI"
#define ERR_INVALIDURI	"Bad Request - Invalid specification"
#define ERR_NOMESSAGE	"Bad Request - No such message"
#define ERR_NOBOUNDARY	"Bad Request - Can't find the MIME boundary"
#define ERR_NOFILES	"Bad Request - No files are attached to this message"
#define ERR_NOFILE	"Bad Request - File not found"
#define ERR_NOFRAMES	"ywcmail requires frames"
#define LOGIN_TITLE	"ywcmail / 0.95 Login"
#define LOGIN_USERNAME	"Login Name"
#define LOGIN_PASSWORD	"Password"
#define LOGIN_POP3HOST	"POP3 Server"
#define LOGIN_SMTPHOST	"SMTP Server"
#define LOGOUT_TITLE	"ywcmail 0.95 Logout"
#define LOGOUT_MESSAGE \
"You have successfully logout.<BR>\n"
#define YWC_TOP_INBOX	"INBOX"
#define YWC_TOP_COMPOSE	"COMPOSE"
#define YWC_TOP_LOGOUT	"LOG OUT"
#define YWC_NOMAIL	"You have no messages in your mailbox."
#define YWC_PREVIOUS	"Previous"
#define YWC_NEXT	"Next"
#define YWC_FROM	"From"
#define YWC_TO		"To"
#define YWC_SUBJECT	"Subject"
#define YWC_DATE	"Date"
#define YWC_SIZE	"Size"
#define YWC_SELECTALL	"Select all messages"
#define YWC_DELSELECTED	"Delete selected messages"
#define YWC_REPLY	"Reply"
#define YWC_FORWARD	"Forward"
#define YWC_DELETE	"Delete"
#define YWC_VIEW_SRC	"View source"
#define YWC_ATTACHMENTS	"Attachments"
#define YWC_CC		"CC"
#define YWC_BCC		"BCC"
#define YWC_FILE	"File"
#define YWC_REPLYLINE	"--- %s wrote:\n"
#define YWC_FWD_FROM	"> From:    %s\n"
#define YWC_FWD_SUBJECT	"> Subject: %s\n"
#define YWC_FWD_DATE	"> Date:    %s\n"
#define YWC_SENDMAIL	"Send Mail"
#define YWC_SENDINGFILE	"Sending file '%s' (%d bytes)<BR>\n"
#define YWC_DELETING	"Deleting message %d..."
#define YWC_DELETE_OK	"success."
#define YWC_DELETE_BAD	"failure."
#define YWC_MIME	"This is a multi-part message in MIME format."
#define YWC_SMTP_FROMOK	"Sender '%s' OK.<BR>\n"
#define YWC_SMTP_RCPTOK	"Recipient '%s' OK.<BR>\n"
#define YWC_SMTP_SENT	"Message Sent."


config.h
/*
    ywcmail -- simple pop3/smtp web e-mail cgi
    Copyright (C) 2002 Chan Yin Chi < 0330@bigfoot.com >

*/
#define YWC_SERVERNAME	" ywcmail / 0.95"

/*
 * APPEND_ADDRESS will let you override the hostname part of return e-mail
 * addresses.  If you don't need it, then don't use it.  You probably will
 * want to use this if the return address domain name is more than two levels.
 * If the username is the full return address (ie, virtual mail systems),
 * you may want to define APPEND_ADDRESS as a blank string.  If you don't use
 * APPEND_ADDRESS, your return address will be the top two levels of the pop3
 * host's name.
 */
#define APPEND_ADDRESS	"@yingwa.edu.hk"

/*
 * If #defined, NO_USER_HOSTS will remove the user's ability to choose their
 * own host names.  Be sure to set POP3_HOST and SMTP_HOST correctly if you
 * choose to use this option.
 */
#define NO_USER_HOSTS

/*
 * BASE_IMAGE_URL tells the cgi where the images are.  STYLESHEET can be used
 * to point to any CSS file you may want to use.
 */
#define BASE_IMAGE_URL	"/webmail/"
#define STYLESHEET	"/webmail/style.css"

/*
 * These colour #defines allow easy customization of may of the colours used
 * in the CGI.  Some of these colours may be overridded if a stylesheet is in
 * use (i.e. COLOR_TRIM and COLOR_TRIMTEXT).
 */
#define COLOR_EDITFORM	"#E0E0E0"
#define COLOR_FTEXT	"#F0F0F0"
#define COLOR_LINKS	"#0000FF"
#define COLOR_TRIM	"#000050"
#define COLOR_TRIMTEXT	"#FFFFFF"

/*
 * If there are more than MAX_LIST_SIZE messages, the message list will be
 * broken up into multiple pages.
 */
#define MAX_LIST_SIZE	20

/*
 * This setting is not an inactivity timer, but instead limits the absolute
 * maximum time in seconds that the cgi is allowed to run.
 */
#define MAX_RUNTIME	1800
/*
 * POP3_HOST and SMTP_HOST really should be changed to reflect the fully
 * qualified domain names of the mail servers.  The port settings typically
 * shouldn't need to be changed.
 */
#define POP3_HOST	"localhost"
#define SMTP_HOST	"localhost"
#define IMAP_PORT	143
#define POP3_PORT	110
#define SMTP_PORT	25

/*
 * You MUST support either POP3, IMAP, or both, otherwise the CGI doesn't work.
 * IMAP support is still highly experimental right now, so I recommend against
 * using it.
 */
//#define SUPPORT_IMAP
#define SUPPORT_POP3

/*
 * SMTP_AUTH enables plain auth with smtp servers using the pop3/imap username
 * and password.
 */
//#define SMTP_AUTH

/*
 * MAX_POSTSIZE limits the total size of allowable POSTS, and as a result can
 * be used to control the maximum size of a message and/or attachments.
 */
#define MAX_POSTSIZE	50554432 /* 50 MB limit for allowed POST sizes */

#define RAW_MESSAGES
#define USE_FRAMES
#define USE_SYSLOG

/*
 * If you want to change the language, uncomment the appropriate strings header
 * file here.  #include one, and only one strings file.
 */
 #include "i18n/strings-en.h"	// English
// #include "i18n/strings-zh_TW.h"	// Traditional Chinese
// # include "i18n/strings-zh_CN.h"     // Simplified Chinese

//Users allowed to access yingwa mail
#define ALLOWED_USERS "/usr/local/yingwa/allow"

#define LOG_FILENAME "/usr/local/apache/htdocs/webmail_main_log.txt"

